components/layout/query.rs

```diff
@@ -864,3 +876,173 @@ pub fn process_style_query<N: LayoutNode>(requested_node: N)
+    for item in results {
+        match item {
+            InnerTextItem::Text(s) => {
+                if max_req_line_break_count > 0 {
+                    // Step 5.
+                    for _ in 0..max_req_line_break_count {
+                        inner_text.push("\u{000A}".to_owned());
+                    }
+                    max_req_line_break_count = 0;
+                }
+                // Step 3.
+                if !s.is_empty() {
+                    inner_text.push(s.to_owned());
+                }
+            },
+            InnerTextItem::RequiredLineBreakCount(count) => {
+                // Step 4.
+                if inner_text.len() == 0 {
+                    // Remove required line break count at the start.
+                    continue;
+                }
+                // Store the count if it's the max of this run,
+                // but it may be ignored if no text item is found afterwards,
+                // which means that these are consecutive line breaks at the end.
+                if count > max_req_line_break_count {
+                    max_req_line_break_count = count;
+                }
+            }
+        }
+    }
+    inner_text.into_iter().collect()
+}
+
+// https://html.spec.whatwg.org/multipage/#inner-text-collection-steps
+#[allow(unsafe_code)]
+fn inner_text_collection_steps<N: LayoutNode>(node: N,
+                                              display_list: &Option<Arc<DisplayList>>,
+                                              results: &mut Vec<InnerTextItem>) {
+    if !display_list.is_some() {
+        warn!("We should have a display list at this point. Cannot get inner text");
+        return;
+    }
+
+    let mut items = Vec::new();
+    for child in node.traverse_preorder() {
+        let node = match child.type_id() {
+            LayoutNodeType::Text => {
+                child.parent_node().unwrap()
+            },
+            _ => child,
+        };
+
+        let element_data = unsafe {
+            node.get_style_and_layout_data().map(|d| {
+                &(*(d.ptr.as_ptr() as *mut StyleData)).element_data
+            })
+        };
+
+        if element_data.is_none() {
+            continue;
+        }
+
+        let style = match element_data.unwrap().borrow().styles.get_primary() {
+            None => continue,
+            Some(style) => style.clone(),
+        };
+
+        // Step 2.
+        if style.get_inheritedbox().visibility != Visibility::Visible {
+            continue;
+        }
+
+        // Step 3.
+        let display = style.get_box().display;
+        if !child.is_in_document() || display == Display::None {
+            continue;
+        }
+
+        match child.type_id() {
+            LayoutNodeType::Text => {
+                // Step 4.
+                let opaque_child = child.opaque();
+                for item in &display_list.as_ref().unwrap().list {
+                    if item.base().metadata.node != opaque_child {
```

Comment:
This looks... not great, this traverses the whole display list for the page for each text-node, right?